singularity-eos/eos/eos_mgusup.hpp

```diff
@@ -0,0 +1,365 @@
+  }
+  PORTABLE_INLINE_FUNCTION Real GruneisenParamFromDensityInternalEnergy(
+      const Real rho, const Real sie, Real *lambda = nullptr) const {
+    return robust::ratio(_G0 * _rho0, rho);
+  }
+  PORTABLE_INLINE_FUNCTION void FillEos(Real &rho, Real &temp, Real &energy, Real &press,
+                                        Real &cv, Real &bmod, const unsigned long output,
+                                        Real *lambda = nullptr) const;
+  PORTABLE_INLINE_FUNCTION
+  void ValuesAtReferenceState(Real &rho, Real &temp, Real &sie, Real &press, Real &cv,
+                              Real &bmod, Real &dpde, Real &dvdt,
+                              Real *lambda = nullptr) const;
+  // Generic functions provided by the base class. These contain e.g. the vector
+  // overloads that use the scalar versions declared here
+  SG_ADD_BASE_CLASS_USINGS(MGUsup)
+  PORTABLE_INLINE_FUNCTION
+  int nlambda() const noexcept { return 0; }
+  static constexpr unsigned long PreferredInput() { return _preferred_input; }
+  static inline unsigned long scratch_size(std::string method, unsigned int nelements) {
+    return 0;
+  }
+  static inline unsigned long max_scratch_size(unsigned int nelements) { return 0; }
+  PORTABLE_INLINE_FUNCTION void PrintParams() const {
+    static constexpr char st[]{"MGUsup Params: "};
+    printf("%s rho0:%e T0:%e Cs:%e s:%e\n  G0:%e Cv0:%e E0:%e S0:%e\n", st, _rho0, _T0,
+           _Cs, _s, _G0, _Cv0, _E0, _S0);
+    printf("\n\n");
+  }
+  // Density/Energy from P/T not unique, if used will give error
+  PORTABLE_INLINE_FUNCTION void
+  DensityEnergyFromPressureTemperature(const Real press, const Real temp, Real *lambda,
+                                       Real &rho, Real &sie) const;
+  inline void Finalize() {}
+  static std::string EosType() { return std::string("MGUsup"); }
+  static std::string EosPyType() { return EosType(); }
+
+ private:
+  static constexpr const unsigned long _preferred_input =
+      thermalqs::density | thermalqs::specific_internal_energy;
+  Real _rho0, _T0, _Cs, _s, _G0, _Cv0, _E0, _S0;
+  void CheckMGUsup();
+};
+
+inline void MGUsup::CheckMGUsup() {
+
+  if (_rho0 < 0.0) {
+    PORTABLE_ALWAYS_THROW_OR_ABORT("Required MGUsup model parameter rho0 < 0");
+  }
+  if (_T0 < 0.0) {
+    PORTABLE_ALWAYS_THROW_OR_ABORT("Required MGUsup model parameter T0 < 0");
+  }
+  if (_Cs < 0.0) {
+    PORTABLE_ALWAYS_THROW_OR_ABORT("Required MGUsup model parameter Cs < 0");
+  }
+  if (_s < 0.0) {
+    PORTABLE_ALWAYS_THROW_OR_ABORT("Required MGUsup model parameter s < 0");
+  }
+  if (_G0 < 0.0) {
+    PORTABLE_ALWAYS_THROW_OR_ABORT("Required MGUsup model parameter G0 < 0");
+  }
+  if (_Cv0 < 0.0) {
+    PORTABLE_ALWAYS_THROW_OR_ABORT("Required MGUsup model parameter Cv0 < 0");
+  }
+}
+
+PORTABLE_INLINE_FUNCTION Real MGUsup::HugPressureFromDensity(Real rho) const {
+  Real eta = 1.0 - robust::ratio(_rho0, rho);
+  return _Cs * _Cs * _rho0 * robust::ratio(eta, (1.0 - _s * eta) * (1.0 - _s * eta));
```

Comment:
Would it make sense to enforce a maximum compressibility here instead?

Reply:
So, I think I warn and throw the code if 1-s \eta gets negative. I think I even point out that the s and \eta combined is non-physical.

Reply:
I have put in this comment where it will break the code, that is, in HugTemperatureFromDensity. It would definitely break TBulkModulusFromDensityTemperature but not necessarily HugPressureFromDensity.
Let me think about what is the correct thing to do here, from physics and thermodynamics.

Reply:
> So, I think I warn and throw the code if 1-s \eta gets negative. I think I even point out that the s and \eta combined is non-physical.

Ah I see you thrown in the `HugTemperatureFromDensity()` function, but unless I'm misunderstanding the code flow, I don't think this would be triggered for e.g. `PressureFromDensityInternalEnergy()`.

> Let me think about what is the correct thing to do here, from physics and thermodynamics.

Sounds good 👍 .

Our solution in the other Gruneisen EOS where we just impose a maximum was mainly an attempt to mirror Pagosa's implementation. However, we gave thought to other approaches like returning the infinity as the appropriate limit, but we decided against approaches like that since they were more tricky from a numerical procedure. With our approach, we just selected a maximum density just before the singularity in the hope that the condition would be self-correcting (i.e. such a high pressure would immediately cause the material to resist further compression).

Regardless though, our approach isn't perfect and may benefit from improvement. I'll look forward to your future thoughts.

Reply:
@jonahm-LANL feel free to correct my memory of the prior density limit discussion (from two years ago?)

Reply:
It also doesn't have to be consistent between this one and the Pagosa-based one. So long as we document what we do I think any reasonable choice is fine.

Reply:
> @jonahm-LANL feel free to correct my memory of the prior density limit discussion (from two years ago?)

I think you remember correctly.

Reply:
I resolve this but continue to think about it and discuss on Mattermost